### ISSUE TITLE
Feature: Allow application definition define hierarchy of widgets

### ISSUE THREAD

**Issue by kitsonk** [dojo/app#30]
# Enhancement

We could enhance the application definition to semantics so that it could behave similar to the way the markup can describe a hierarchy/structure of an application.  In that the `dojo/app` instance would be able to instantiate the widget instances as well as organise them into their child/parent relationships.

Parent Widgets ideally manage their children from state, so once instantiated, the `dojo/app` would simply need to set the children to an array or object hash of the children.

Ideally the `dojo/app` would also detect if the `children` property of the widget is a `List` or `Map` and set the state as either an array or object literal.

Since the children widgets need to be instantiated, it may not make sense to pass them in the `options` of the widget descriptor, but instead introduce a optional property of `children` which can then take an array of object map of IDs or widget descriptors.  For examples:

``` typescript
app.loadDefinition({
    widgets: [
        {
            id: 'foo',
            factory: createParentListWidget,
            children: [
                'bar',
                { id: 'baz', factory: createWidget }
            ]
        }, {
            id: 'qat',
            factory: createParentMapWidget,
            children: {
                'child1': 'qux',
                'child2': { id: 'quux', factory: createWidget }
            }
        }
    ]
});
```

**Comment by novemberborn** [comment]
> Parent Widgets ideally manage their children from state, so once instantiated, the dojo/app would simply need to set the children to an array or object hash of the children.

You mean doing `instance.children = ['bar', 'baz']`?

Should the app resolve the widgets before instantiating the parent?

> Ideally the dojo/app would also detect if the children property of the widget is a List or Map and set the state as either an array or object literal.

Seems easier if dojo/app takes whatever was specified, and the specific widgets throw if the wrong type is assigned?

---

Note that the app still won't be able to automatically attach widgets. Perhaps an `app.attach(element, widgetID[])` sort of method would suffice?

**Comment by kitsonk** [comment]
This relates to #33, but we should keep it open and make sure we effectively address it.

**Comment by novemberborn** [comment]
https://github.com/dojo/app/pull/38 implements defining children in widget definitions. It does not provide an automatic way of rendering the root widgets.

**Comment by novemberborn** [comment]
@kitsonk what do you want to do with this? I can rebase #38 if that helps.

**Comment by kitsonk** [comment]
We should put this on hold until we can have another deep dive on `app`.

**Comment by novemberborn** [comment]
We'll allow widget definitions to have an optional `root` property. Roots can be `Element`s or query selector `string`s. When the app is started, widgets with roots are automatically appended to the DOM.

Need to investigate whether roots must still be contained in `app-projector` elements, or if each root is automatically wrapped in its own projector (even if there is a parent projector in the DOM tree).

@kitsonk suggested allowing widget IDs to be specified as roots. Perhaps we can revisit that at a later stage — the `children` definitions should suffice?

**Comment by novemberborn** [comment]
What should happen if the widget root does not exist? Ignore seems good and would work better with incremental renders (#84).

**Comment by novemberborn** [comment]
> Need to investigate whether roots must still be contained in app-projector elements, or if each root is automatically wrapped in its own projector (even if there is a parent projector in the DOM tree).

Realized that projectors cannot be nested. So create one if necessary, otherwise use the existing projector.

**Comment by novemberborn** [comment]
Given discussion in https://github.com/dojo/app/issues/84#issuecomment-255359483 we should only render rooted widgets once, when the app is started. That means building on top of https://github.com/dojo/app/pull/76 (and ensuring `start()` can only be called once).

Selectors will be relative to the `root` option provided to `start()`. Elements must be contained within the `root`.

**Comment by novemberborn** [comment]
I'm not sure whether this is feasible at all. Take this snippet from TodoMVC:

```html
<section class="todoapp">
  <header>
    <app-projector>
      <dojo-widget data-uid="title" data-options='{"tagName": "h1"}'></dojo-widget>
      <div is="app-widget" id="new-todo"/>
    </app-projector>
  </header>
  <app-projector>
    <div is="app-widget" id="main-section"></div>
    <div is="app-widget" id="todo-footer"></div>
  </app-projector>
</section>
```

Without `<app-projector>` elements the selectors would be:

|uid|selector|
|---|---|
|`title`|`.todoapp > header`|
|`new-todo`|`.todoapp > header`|
|`main-section`|`.todoapp`|
|`todo-footer`|`.todoapp`|

However it'd be hard to write logic that realizes these widgets could *share* a projector, so more likely they each get their own. So instead the projectors have to be added to the DOM and included in the selectors:

|uid|selector|
|---|---|
|`title`|`.todoapp > header > app-projector`|
|`new-todo`|`.todoapp > header > app-projector`|
|`main-section`|`.todoapp > app-projector`|
|`todo-footer`|`.todoapp > app-projector`|

That would work, but begs the question in which order the widgets should be appended. Given that we want to support backends generating a POJO, just relying on registration order seems troublesome.

So then these sibling widgets should probably be created as container children, with that container having the `.todoapp > header` or `.todoapp` selectors. Or alternatively the projectors can somehow be declared and identified in the POJO.

Using containers nor declaring projectors deals with scenarios where the widgets need to be inserted before or after particular siblings. So that would require further arguments in the POJO.

At which point, why not create a projector use hyperscript using an `afterRealize` callback in `App#start()`? That'd be way easier than trying to declare these relationships in a POJO. Or, indeed, use the declarative approach.

Oh, and all this would only cover the initial render. As stated in https://github.com/dojo/app/issues/84#issuecomment-255359483 we need something more advanced for incremental rendering. Quite how this POJO rooting scheme would fit into that I do not know.

---

**tl;dr** automatically rendering widgets based on their definitions only works in the simplest of scenarios. The functionality is too confusing as soon as more than one widget is involved, so we shouldn't bother building it at all.

**Comment by kitsonk** [comment]
We will revisit under other guises.
